### DISCUSSION TITLE
Publish="false" only works on PackageReference, not ProjectReference

### DISCUSSION THREAD

**Discussion by Mrcubix** [dotnet/runtime#108159] [Q&A]
Currently developing extensions for a driver.

Some of the core dependencies are already loaded in the default context at runtime, which means there isn't a need to package them with my extension as the app will provide them (plugins are built for a range versions of the app).

I can exclude dlls from from a package using `Publish="false"` like such :
```csproj
<ItemGroup>
    <PackageReference Include="MyApp.Plugin" Version="1.0.0" Publish="false" />
</ItemGroup>
```
However, attempting to use the same property for a project reference (`ProjectReference`) will result in the project's files being copied to the output folder on publish :
```csproj
<ItemGroup>
    <ProjectReference Include="../.modules/MyApp/MyApp.Plugin/MyApp.Plugin.csproj" />
</ItemGroup>
```
(in my use case, I am using a git submodule as some of the components I need are not available as nuget packages).

Since this isn't working, I'm wondering if there is another way to prevent those files from being copied to the publish folder, that wouldn't involve a post build target, or a cleanup step in my build script.

Currently, I'm building my extensions to a temporary folder where i then proceed to move specific files to the final build folder, not ideal, and prone to break if not carefully done.

Any Suggestions?

## Update 1

`Private="False"` Seem to only work for the dlls generated directly by the target project, it does not prevent dlls from its dependencies from being included in the publish folder.

`DisableTransitiveProjectReferences="true"` could partially solve that issue, but then i would have to add a project reference for all subdependencies i might need.

**Comment by KeterSCP** [comment]
Maybe what you are looking for is `Private="false"`: https://learn.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022#projectreference

**Reply by Mrcubix** [reply]
Already tried, it also does not seem to do anything.

**Reply by KeterSCP** [reply]
Are you sure you specified `Private="false"` and not `Private="true"`? Because when I test this locally with `ProjectA`, referencing `ProjectB`, and having following ProjectA.csproj:

```xml
<ItemGroup>
    <ProjectReference Include="../ProjectB.csproj" Private="false" />
</ItemGroup>
```

After calling `dotnet publish`, in the `publish` folder there are only DLLs from `ProjectA`

**Reply by Mrcubix** [reply]
Afaik, i tried both options, along with PrivateAssets, ReferenceOutputAssembly (which break compilation), Content, None, CopyLocal.

**Reply by Mrcubix** [reply]
Turns out it works, if the reference doesn't rely on other dependencies, otherwise, those will still be included.
In this case, `MyApp.Plugin` will be excluded, but its dependencies will still be included in the publish folder, which i don't want.

The `DisableTransitiveProjectReferences="true"` could partially solve that issue, but then i would have to add a project reference for all subdependencies i might need.

Ideally, i would need a recursive alternative to `Private="False"`, which `Publish="False"` is for `PackageReference`

**Comment by huoyaoyuan** [comment]
See https://github.com/dotnet/sdk/issues/41529 and https://github.com/dotnet/sdk/issues/31636

**Reply by Mrcubix** [reply]
Already tried simillar solutions using the same same properties yesterday.

- Private=False does nothing,
- ExcludeAssets="runtime" does nothing,
- DisableTransitiveProjectReferences=true breaks the build process.

All 3 within the same reference does not seem to do anything either.

I remember reading an issue that was open since 2017, where the same issue I'm encountering has been reported.
Almost 7 years later, it's unfortunately still not fixed. (https://github.com/dotnet/sdk/issues/952#issuecomment-373854501)
